### PR TITLE
 Job builder

### PR DESCRIPTION
**Pull request by rustworthy**
As per [feature request](https://github.com/jonhoo/faktory-rs/issues/9)

- - -
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/faktory-rs/37)

### CODE REVIEW

File: src/bin/loadtest.rs
Comment (jonhoo) [inline src/bin/loadtest.rs:74]: This feels like it belongs in a code block in the documentation. The `loadtest` binary isn't really intended as user-facing documentation, but rather as a stand-lone tool.

File: src/error.rs
Comment (jonhoo) [inline src/error.rs:25]: nit: please preserve the empty line between each enum variant

File: src/proto/single/mod.rs
Comment (jonhoo) [inline src/proto/single/mod.rs:165]: This should probably continue to not be `PartialEq`.

File: src/error.rs
Comment (jonhoo) [inline src/error.rs:98]: Do we ever expect to have other client-side errors than malformed jobs? I wonder if we should skip the intermediate `enum` here and instead have `MalformedJob` be a variant directly on `Error`.

File: src/proto/single/mod.rs
Comment (jonhoo) [inline src/proto/single/mod.rs:42]:
```suggestion
/// let _job = JobBuilder::default()
///     .kind("order")
///     .args(vec!["ISBN-13:9781718501850"])
///     .build()?;
/// # Ok(())
```

File: src/proto/single/mod.rs
Comment (jonhoo) [inline src/proto/single/mod.rs:62]: Given that `kind` is a required argument, I'd rather it have to be passed to `builder`. That seems to be the generally-recommended pattern for options that must be set for the build to succeed. Look at the doc code elsewhere, that doesn't seem to be the case at the moment?

File: src/proto/single/mod.rs
Comment (jonhoo) [inline src/proto/single/mod.rs:187]: To avoid duplicating code here, can we now make this just call `JobBuilder::new(kind).args(args).build()`?

File: src/proto/single/mod.rs
Comment (jonhoo) [reply src/proto/single/mod.rs:62]: An alternative if the derive crate doesn't let you do that is to have it generate a private constructor (i.e., to make `JobBuilder` not implement `Default` ), and instead have a `Job::builder(kind)` that returns a `JobBuilder` where the `kind` has been set.

File: src/proto/single/mod.rs
Comment (jonhoo) [inline src/proto/single/mod.rs:150]: If this is the only kind of misconfiguration there can be for a job, I'm inclined to not have it return an error (and thus get rid of the new error type and the `Result` in `build`'s return type. Especially because `priority` is already a `pub` field, so users can always set it to the wrong value elsewhere.

**Review by jonhoo — CHANGES_REQUESTED** [review]
This looks like a good starting point, thanks for taking the time! Left some comments.

File: src/proto/single/mod.rs
Comment (rustworthy) [reply src/proto/single/mod.rs:62]: This 'default' in current implementation is definitely deceptive: as if it were enough to build a simple job, while it's not and it really should be followed/enriched by `.kind()`  and `.args()` , otherwise the error will be returned.

With these updates:

1)  as per your suggestion, the required arg is set of the JobBuilder right away (either via JobBuilder::new("jobtype" or  - also added as an alternative way to instantiate a builder - via Job::builder("jobtype")));

2) the 'args' field has been given a default of 'Vec::new()' on the JobBuilder;

3) as per your suggestions, we got rid of the 'priority' validation (As a side-note: on the second thought, I realized that adding a 'JobPriority' (e.g. JobPriority::from(3)) new type would have been a better solution compared to erroring back at a job's build time. In the new type case they would be able to set/reset the priority as many time as they want and it would still be valid. But this 1) looks like a subject of a different PR and 2) not necessary in the first place);

We can:
1. Create a new Job with JobBuilder infallibly;

2. Get rid of the 'Client' error variant (I am using this one in another branch on tracking job's progress for Faktory enterprise, BUT I am using it there in the exact same case: checking that 'percent' value in the range from 0 to 100. I now think I should be adding this validation there to not constrain the user and plus no instructions on this are given in Faktory docs).

File: src/proto/single/mod.rs
Comment (rustworthy) [reply src/proto/single/mod.rs:150]: Got rid of this.

File: src/proto/single/mod.rs
Comment (rustworthy) [reply src/proto/single/mod.rs:42]: Added usage examples taking into account that the build is not returning result any longer

File: src/error.rs
Comment (rustworthy) [reply src/error.rs:98]: Looks like with other suggestions of yours applied, this variant is no longer needed (for now :))

File: src/bin/loadtest.rs
Comment (rustworthy) [reply src/bin/loadtest.rs:74]: restored

**Review by jonhoo — CHANGES_REQUESTED** [review]

**Review by jonhoo — APPROVED** [review]

**Comment by jonhoo** [comment]
I'll hold off on releasing this until #34 lands.

**Comment by jonhoo** [comment]
Released in 0.12.2 :tada:

**Pull request merged**
